File: Assets/PythonScripts/objdem.py
#!/usr/bin/env python

# Kevin Connors 2017

import sys, os, time, array, numpy, math, utm
from urllib.request import urlopen
from urllib.error import HTTPError
from scipy.spatial import Delaunay

# the database
worldwind = 'https://data.worldwind.arc.nasa.gov'
m_per_deg_lat = 111619.0

elevation_data = []
mean_x = 0
mean_y = 0
min_z = 0


def fetch_elevation_data(min_long, min_lat, max_long, max_lat, resolution):
    if (resolution < 30):
        resolution = 30

    resolution_in_deg = resolution / m_per_deg_lat

    long_range = max_long - min_long
    lat_range = max_lat - min_lat

    width = round(long_range / resolution_in_deg)
    height = round(lat_range / resolution_in_deg)

    print("    Querying database...")

    res = urlopen(worldwind +
                  '/elev?'
                  'service=WMS'
                  '&request=GetMap'
                  '&layers=NED'
                  '&crs=EPSG:4326'
                  '&format=image/bil'
                  '&transparent=FALSE'
                  '&width=' + str(width) +
                  '&height=' + str(height) +
                  '&bgcolor=0xFFFFFF'
                  '&bbox=' + str(min_long) + ',' + str(min_lat) + ',' + str(max_long) + ',' + str(max_lat) +
                  '&styles='
                  '&version=1.3.0')

    print("    Converting data...")
    f = open('data.bil', 'wb')
    f.write(res.read())
    f.close()

    # Read from file
    b = array.array("h")
    with open("data.bil", "rb") as f:
        b.fromfile(f, width * height)
    if sys.byteorder == "big":
        b.byteswap()

    for x in range(0, height):
        row = []
        for y in range(0, width):
            start = width * x
            row.append(b[start + y])
        elevation_data.append(row)

    print("    Fetched elevation data successfully.")


# end function


def fetch_image_data(min_long, min_lat, max_long, max_lat, resolution, filename="map.tiff"):
    if (resolution < 30):
        resolution = 30

    if min_lat < -60.0 or max_lat > 84.0:
        print("Landsat data is not available for values of latitude outside of the range -60.0 to 84.0")
        # TODO find a database to serve this data
        return

    resolution_in_deg = resolution / m_per_deg_lat

    long_range = max_long - min_long
    lat_range = max_lat - min_lat

    width = round(long_range / resolution_in_deg)
    height = round(lat_range / resolution_in_deg)

    print("    Querying database...")
    res = urlopen(worldwind +
                  '/landsat?'
                  'service=WMS'
                  '&request=GetMap'
                  '&layers=esat'
                  '&crs=EPSG:4326'
                  '&format=image/tiff'
                  '&transparent=FALSE'
                  '&width=' + str(width) +
                  '&height=' + str(height) +
                  '&bgcolor=0xFFFFFF'
                  '&bbox=' + str(min_long) + ',' + str(min_lat) + ',' + str(max_long) + ',' + str(max_lat) +
                  '&styles='
                  '&version=1.3.0')

    f = open(filename, 'wb')
    f.write(res.read())
    f.close()

    print("    Image created successfully.")


# end function

def elevation_points_to_xyz(min_long, min_lat, max_long, max_lat, resolution):
    global mean_x
    global mean_y
    global min_z

    resolution_in_deg = resolution / m_per_deg_lat

    data = []

    print("    Converting points to UTM...")
    for i in range(0, len(elevation_data)):
        for j in range(0, len(elevation_data[0])):
            long = min_long + resolution_in_deg * j
            lat = max_lat - resolution_in_deg * i
            (x, y, _zone_num, _zone_letter) = utm.from_latlon(lat, long)
            z = elevation_data[i][j]
            element = [round(x), round(y), z]
            data.append(element)

    xs = list(map(lambda e: e[0], data))
    ys = list(map(lambda e: e[1], data))
    zs = list(map(lambda e: e[2], data))

    mean_x = math.floor((min(xs) + max(xs)) / 2)
    mean_y = math.floor((min(ys) + max(ys)) / 2)
    min_z = math.floor(min(zs))

    data.append([min(xs) - 10, min(ys) - 10, min_z - 1])
    data.append([min(xs) - 10, max(ys) + 10, min_z - 1])
    data.append([max(xs) + 10, min(ys) - 10, min_z - 1])
    data.append([max(xs) + 10, max(ys) + 10, min_z - 1])

    return list(map(lambda e: [e[0] - mean_x, e[1] - mean_y, e[2] - min_z], data))
# end function


# precondition: fetch_elevation_data has been called
def write_points_to_obj(min_long, min_lat, max_long, max_lat, resolution, filename="dem.obj"):
    try:
        os.remove(filename)
        f = open(filename, 'w')
    except FileNotFoundError:
        f = open(filename, 'w')

    points = elevation_points_to_xyz(min_long, min_lat, max_long, max_lat, resolution)

    print("    Writing points...")
    # write vertices
    for point in points:
        f.write("v " + str(point[0]) + " " + str(point[1]) + " " + str(point[2]) + '\n')

    # write vertex textures for uv mapping
    height = len(elevation_data)
    width = len(elevation_data[0])

    print("    UV mapping...")
    for i in range(0, height):
        for j in range(0, width):
            f.write("vt " + str(j / width) + " " + str(height - (i / height)) + ' 0\n')

    xy_points = numpy.array(list(map(lambda x: [x[0], x[1]], points)))

    tris = Delaunay(xy_points)

    a = len(points) - 1
    b = len(points) - 2
    c = len(points) - 3
    d = len(points) - 4

    print("    Writing triangles...")
    # write facets
    for simplex in tris.simplices:
        # don't compute for (0,0) or (width,height)
        if (simplex[0] != a and
                    simplex[1] != a and
                    simplex[2] != a and
                    simplex[0] != b and
                    simplex[1] != b and
                    simplex[2] != b and
                    simplex[0] != c and
                    simplex[1] != c and
                    simplex[2] != c and
                    simplex[0] != d and
                    simplex[1] != d and
                    simplex[2] != d):
            f.write("f " + str(simplex[0] + 1) + "/" + str(simplex[0] + 1) + " "
                    + str(simplex[1] + 1) + "/" + str(simplex[1] + 1) + " "
                    + str(simplex[2] + 1) + "/" + str(simplex[2] + 1) + '\n')

    f.close()
# end function


def find_centroid_from_obj(file, zone_number, zone_letter):
    xyz = []

    with open(file) as f:
        lines = f.readlines()
        for line in lines:
            if line.find("v ") == 0:
                content = line.strip('\n').strip('v').strip().split()
                xyz.append(content)

    xs = list(map(lambda x: float(x[0]), xyz))
    ys = list(map(lambda x: float(x[1]), xyz))

    centerX = max(xs) - ((max(xs) - min(xs)) / 2)
    centerY = max(ys) - ((max(ys) - min(ys)) / 2)

    centerLat, centerLong = utm.to_latlon(centerX, centerY, zone_number, zone_letter)

    return (centerLong, centerLat)
# end function


def scale_down_obj(file):
    try:
        i = file.index(".obj")
    except ValueError:
        print("Filename must end in .obj")
        return

    outfile = file[0:i]
    fo = open(outfile + "Scaled.obj", "w")

    xyz_points = []

    f = open(file, "r")
    lines = f.readlines()

    for line in lines:
        if line.find("v ") == 0:
            line = line.strip('\n').strip('v').strip().split()
            xyz_points.append([float(line[0]), float(line[1]), float(line[2])])

    xyz_points = list(map(lambda e: [e[0] - mean_x, e[1] - mean_y, e[2] - min_z], xyz_points))

    f.seek(0)
    p = 0
    for line in lines:
        if line.find("v ") == 0:
            line = "v " + str(xyz_points[p][0]) + " " + str(xyz_points[p][1]) + " " + str(xyz_points[p][2]) + "\n"
            fo.write(line)
            p += 1
        else:
            fo.write(line)
# end function


def main():
    if len(sys.argv) == 10 and sys.argv[1] == "photogrammetry":

        print("Constructing a DEM around photogrammetry model...")

        long_range = float(sys.argv[2])
        lat_range = float(sys.argv[3])
        resolution = float(sys.argv[4])
        zone_number = float(sys.argv[5])
        zone_letter = sys.argv[6]
        photogrammetry_filename = sys.argv[7]
        model_filename = sys.argv[8]
        image_filename = sys.argv[9]

        print("    Finding centroid...")
        (centroidLong, centroidLat) = find_centroid_from_obj(photogrammetry_filename, zone_number, zone_letter)

        min_long = centroidLong - long_range / 2
        min_lat = centroidLat - lat_range / 2
        max_long = centroidLong + long_range / 2
        max_lat = centroidLat + lat_range / 2

    elif len(sys.argv) == 2 and sys.argv[1] == 'default':
        min_long = -79.65
        min_lat = 37.6
        max_long = -79.35
        max_lat = 37.9
        resolution = 90
        model_filename = "dem.obj"
        image_filename = "map.tiff"
    elif len(sys.argv) != 6 and len(sys.argv) != 8:
        print("Invalid number of arguments.  Usage: python objdem.py min_long min_lat max_long max_lat resolution")
        print("Or python objdem.py min_long min_lat max_long max_lat resolution model_filename image_filename")
        return
    else:
        min_long = float(sys.argv[1])
        min_lat = float(sys.argv[2])
        max_long = float(sys.argv[3])
        max_lat = float(sys.argv[4])
        resolution = float(sys.argv[5])

        if len(sys.argv) == 8:
            model_filename = sys.argv[6]
            image_filename = sys.argv[7]

    print("Fetching elevation data...")
    try:
        fetch_elevation_data(min_long, min_lat, max_long, max_lat, resolution)
    except HTTPError:
        print("    Coordinate range out of bounds.  Use a range between 0.01 and 1.")
        print("    Aborting...")
        time.sleep(4)

    os.remove("data.bil")

    print("Fetching image data...")
    try:
        fetch_image_data(min_long, min_lat, max_long, max_lat, 30, filename=image_filename)
    except HTTPError:
        print("    Coordinate range out of bounds.  Use a range between 0.01 and 1.")
        print("    Aborting...")
        time.sleep(4)

    print("Creating .obj file...")
    write_points_to_obj(min_long, min_lat, max_long, max_lat, resolution, filename=model_filename)
    print("Created .obj file successfully.")

    if len(sys.argv) == 10 and sys.argv[1] == "photogrammetry":
        # this must be called after fetching elevation data so that it can be scaled based on the mean_x, mean_y, and min_z values
        print("Scaling photogrammetry .obj down based on elevation data...")
        scale_down_obj(photogrammetry_filename)

    time.sleep(3)
# end function


main()

File: src/Editor/ObjDEM.cs
using System;
using System.Text.RegularExpressions;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using UnityEditor;
using UnityEngine;

public class ObjDEM : ScriptableObject {

    private static string database = "https://data.worldwind.arc.nasa.gov";
    private static float meterPerDegreeLat = 111619.0f;
	private static float baseResolution = 30.0f;
	private static float resolutionInDeg = baseResolution / meterPerDegreeLat;

    private static List<int> elevationData = new List<int>();
    private static int[,] ElevationData;
	public static float meanX = 0.0f;
	public static float meanY = 0.0f;
	public static float minZ = 0.0f;
	public static float xRange = 0.0f;
	public static float yRange = 0.0f;
	public static float zRange = 0.0f;

    private IEnumerator WWWElevationData(string req, int width, int height)
    {
    // from https://forum.unity.com/threads/www-is-not-ready-downloading-yet.131989/
        WWW res = new WWW(req);
        yield return res;
        byte[] data = new byte[res.bytes.Length];
        Array.Copy(res.bytes, data, res.bytes.Length);

        if (BitConverter.IsLittleEndian)
        {
            Array.Reverse(data);
        }

		for (int i = data.Length - 1; i != -1; i -= 2)
        {
            short int16 = (short)(((data[i-1] & 0xFF) << 8) | (data[i] & 0xFF));
			elevationData.Add ((int)int16);
        }

        ElevationData = new int[height, width];

        for (int x = 0; x < height; x++)
        {
           for (int y = 0; y < width; y++)
            {
                int start = width * x;
                ElevationData[x, y] = elevationData[start + y];
            }
        }
    }
    
    // Fetches elevation data from data.worldwind.arc.nasa.gov
    private void FetchElevationData(float minLong, float maxLong, float minLat, float maxLat)
    {
        
        float longRange = maxLong - minLong;
        float latRange = maxLat - minLat;

        int width = (int)Mathf.Round(longRange / resolutionInDeg);
        int height = (int)Mathf.Round(latRange / resolutionInDeg);

        //print("    Querying database...");
        string req = database + "/elev?service=WMS&request=GetMap&layers=mergedSrtm&crs=EPSG:4326&format=image/bil&transparent=FALSE&width=" + width.ToString() + "&height=" + height.ToString() + "&bgcolor=0xFFFFFF&bbox=" + minLong.ToString() + "," + minLat.ToString() + "," + maxLong.ToString() + "," + maxLat.ToString() + "&styles=&version=1.3.0";

        //  from https://stackoverflow.com/questions/33251869/how-to-use-www-in-editor-script
        IEnumerator enumerator = WWWElevationData(req, width, height);

        // Current points to null here, so move it forward
        enumerator.MoveNext();

        // This blocks, but you can always use a thread
        while (!((WWW)(enumerator.Current)).isDone) ;

        // This triggers your 'Debug.Log(www.text)'
        enumerator.MoveNext();

        //print("    Fetched elevation data successfully.")
    
    }

	private void GetElevationDataFromRaw(string rawFile, int numRows, int numCols, bool use16Bit) {

		byte[] data;

		using (FileStream fs = new FileStream (Application.dataPath + "/" + rawFile, FileMode.Open, FileAccess.Read)) {

			data = new byte[fs.Length];

			int numBytesToRead = (int)fs.Length;
			int numBytesRead = 0;
			while (numBytesToRead > 0)
			{
				// Read may return anything from 0 to numBytesToRead.
				int n = fs.Read(data, numBytesRead, numBytesToRead);

				// Break when the end of the file is reached.
				if (n == 0)
					break;

				numBytesRead += n;
				numBytesToRead -= n;
			}
		} 

		if (BitConverter.IsLittleEndian) {
			Array.Reverse(data);
		}

		if (use16Bit) {
			for (int i = data.Length - 1; i > 0; i -= 2) {
				short int16 = (short)(((data [i - 1] & 0xFF) << 8) | (data [i] & 0xFF));
				elevationData.Add ((int)int16);
			}
		} else {
		
			for (int i = data.Length - 1; i != -1; i--) {
				elevationData.Add ((int)data [i]);
			}

		}

		ElevationData = new int[numRows, numCols];

		for (int x = 0; x < numRows; x++)
		{
			for (int y = 0; y < numCols; y++)
			{
				int start = numCols * x;
				ElevationData[x, y] = elevationData[start + y];
			}
		}
	}

    private IEnumerator WWWImageData(string req, string outfile)
    {
        WWW res = new WWW(req);
        yield return res;

        File.WriteAllBytes(Application.dataPath + "/" + outfile, res.bytes);
        AssetDatabase.Refresh();
    }

    // Fetches an image from data.worldwind.arc.nasa.gov
    private void FetchImageData(float minLong, float maxLong, float minLat, float maxLat, string filename)
    {

        if (minLat < -60.0 || maxLat > 84.0)
        {
            throw new Exception("FetchImageDataError: Landsat data is not available for values of latitude outside of the range -60.0 to 84.0");
        }

        float longRange = maxLong - minLong;
        float latRange = maxLat - minLat;

        int width = (int)Mathf.Round(longRange / resolutionInDeg);
        int height = (int)Mathf.Round(latRange / resolutionInDeg);

        string req = database + "/landsat?service=WMS&request=GetMap&layers=esat&crs=EPSG:4326&format=image/tiff&transparent=FALSE&width=" + width.ToString() + "&height=" + height.ToString() + "&bgcolor=0xFFFFFF&bbox=" + minLong.ToString() + "," + minLat.ToString() + "," + maxLong.ToString() + "," + maxLat.ToString() + "&styles=&version=1.3.0";

        IEnumerator enumerator = WWWImageData(req, filename);

        enumerator.MoveNext();

        // This blocks, but you can always use a thread
        while (!((WWW)(enumerator.Current)).isDone) ;

        enumerator.MoveNext();

    }
    
    // converts ElevationData to a Vector3[]
    private List<Vector3> ElevationPointsToXYZ(float minLong, float maxLong, float minLat, float maxLat)
    {

        List<Vector3> data = new List<Vector3>();

        for (int i = 0; i < ElevationData.GetLength(0); i++)
        {
            for (int j = 0; j < ElevationData.GetLength(1); j++) {
                double lon = minLong + resolutionInDeg * j;
                double lat = maxLat - resolutionInDeg * i;
                Vector2 UTM = Helpers.LonLatToUTM(lon, lat);
                float z = (float)ElevationData[i, j];
                Vector3 point = new Vector3(UTM.x, UTM.y, z);
                data.Add(point);
            }
        }

		if (SDTP.SceneFileExists ()) {
			xRange = SDTP.xRange;
			yRange = SDTP.yRange;
			zRange = SDTP.zRange;
			minZ = SDTP.minZ;
			meanX = SDTP.meanX;
			meanY = SDTP.meanY;
		} else {
			IEnumerable<float> xs = data.Select (e => e.x);
			IEnumerable<float> ys = data.Select (e => e.y);
			IEnumerable<float> zs = data.Select (e => e.z);

			float xMin = xs.Aggregate ((a, b) => a < b ? a : b);
			float yMin = ys.Aggregate ((a, b) => a < b ? a : b);
			float zMin = zs.Aggregate ((a, b) => a < b ? a : b);
			float xMax = xs.Aggregate ((a, b) => a > b ? a : b);
			float yMax = ys.Aggregate ((a, b) => a > b ? a : b);
			float zMax = zs.Aggregate ((a, b) => a > b ? a : b);

			// update global variables for later use
			xRange = xMax - xMin;
			yRange = yMax - yMin;
			zRange = zMax - zMin;
			minZ = zMin;
			meanX = (xMin + xMax) / 2;
			meanY = (yMin + yMax) / 2;
		}

        return data.Select(p => new Vector3(p.x - meanX, p.y - meanY, p.z - minZ)).ToList();
    }

	private List<Vector3> ElevationPointsToXYZ(float xOrigin, float yOrigin, float resolution) {

		List<Vector3> data = new List<Vector3>();

		for (int i = 0; i < ElevationData.GetLength(0); i++)
		{
			for (int j = 0; j < ElevationData.GetLength(1); j++) {
				float x = xOrigin + resolution * j;
				float y = yOrigin - resolution * i;
				float z = (float)ElevationData[i, j];
				Vector3 point = new Vector3(x, y, z);
				data.Add(point);
			}
		}

		if (SDTP.SceneFileExists ()) {
			xRange = SDTP.xRange;
			yRange = SDTP.yRange;
			zRange = SDTP.zRange;
			minZ = SDTP.minZ;
			meanX = SDTP.meanX;
			meanY = SDTP.meanY;
		} else {
			IEnumerable<float> xs = data.Select (e => e.x);
			IEnumerable<float> ys = data.Select (e => e.y);
			IEnumerable<float> zs = data.Select (e => e.z);

			float xMin = xs.Aggregate ((a, b) => a < b ? a : b);
			float yMin = ys.Aggregate ((a, b) => a < b ? a : b);
			float zMin = zs.Aggregate ((a, b) => a < b ? a : b);
			float xMax = xs.Aggregate ((a, b) => a > b ? a : b);
			float yMax = ys.Aggregate ((a, b) => a > b ? a : b);
			float zMax = zs.Aggregate ((a, b) => a > b ? a : b);

			// update global variables for later use
			xRange = xMax - xMin;
			yRange = yMax - yMin;
			zRange = zMax - zMin;
			minZ = zMin;
			meanX = (xMin + xMax) / 2;
			meanY = (yMin + yMax) / 2;
		}


		return data.Select(p => new Vector3(p.x - meanX, p.y - meanY, p.z - minZ)).ToList();
	
	}

    // Writes out points to an .obj file
    // Precondition: FetchElevationData has been called
    private void WritePointsToObj(List<Vector3> points, string filename)
    {
        string fp = Application.dataPath + "/" + filename;
        File.Delete(fp);
        File.Create(fp).Dispose();

		List<Vector3> triangles = Helpers.Triangulate(ElevationData.GetLength(1), ElevationData.GetLength(0));

        using (StreamWriter file = new StreamWriter(fp, true))
        {
           
			for (int i = 0; i < points.Count; i++)
			{
				Vector3 p = points[i];
				file.WriteLine("v " + p.x + " " + p.y + " " + p.z);
			}
				
            // write vertex textures for uv mapping
            int height = ElevationData.GetLength(0);
            int width = ElevationData.GetLength(1);

            for (int i = 0; i < height; i++)
            {
                for (int j = 0; j < width; j++)
                {
					file.WriteLine("vt " + (j / (double)width).ToString() + " " + ((height - i) / (double)height).ToString() + " 0");
                }
                
            }

            // write facets
            // don't compute for 0,0 or width,height
            foreach (Vector3 simplex in triangles)
            {
                // cast to int since triangles are in ints, but Vector3 uses floats
                int x = (int)simplex.x;
                int y = (int)simplex.y;
                int z = (int)simplex.z;

                // we add 1 here because the .obj file indexing starts at 1, not 0.
                file.WriteLine("f " + (x + 1).ToString() + "/" + (x + 1).ToString() + " "
                                    + (y + 1).ToString() + "/" + (y + 1).ToString() + " "
                                    + (z + 1).ToString() + "/" + (z + 1).ToString());
            }
        }
        
    }
 
    // returns the long/lat of the centroid of the object
    public static Vector2 FindCentroidOfObj(string filename, string zone)
    {
		if (!filename.EndsWith(".obj")) throw new Exception("ObjDEMError: File must be of type .obj");

		string line;
		List<Vector3> xyzPoints = new List<Vector3>();

		StreamReader file = new StreamReader(Application.dataPath + "/" + filename);

		// read lines matching "v x y z" and turn that into a Vector3, which is put in xyzPoints
		while ((line = file.ReadLine()) != null) {

			if (line.StartsWith("v ")) {
				line = line.Substring (line.IndexOf ("v ") + 2);
				line = line.Trim();
				string[] point = line.Split(new char[] {' '});
				xyzPoints.Add(new Vector3(float.Parse(point[0]), float.Parse(point[1]), float.Parse(point[2])));
			}
		}

		file.Close ();

		IEnumerable<float> xs = xyzPoints.Select(e => e.x);
		IEnumerable<float> ys = xyzPoints.Select(e => e.y);

		float xMin = xs.Aggregate((a, b) => a < b ? a : b);
		float yMin = ys.Aggregate((a, b) => a < b ? a : b);
		float xMax = xs.Aggregate((a, b) => a > b ? a : b);
		float yMax = ys.Aggregate((a, b) => a > b ? a : b);

		float centerX = (xMax + xMin) / 2;
		float centerY = (yMax + yMin) / 2;

		return Helpers.UTMToLatLon (centerX, centerY, zone);
    }
		
    
    public static void ScaleDownObj(string filename)
    {

		if (!filename.EndsWith(".obj")) throw new Exception("ObjDEMError: File must be of type .obj");
			
		string line;
		List<double[]> xyzPoints = new List<double[]>();

		StreamReader file = new StreamReader(Application.dataPath + "/" + filename);

		// read lines matching "v x y z" and turn that into a Vector3, which is put in xyzPoints
		while ((line = file.ReadLine()) != null) {
		
			if (line.StartsWith("v ")) {
				line = line.Substring (line.IndexOf ("v ") + 2);
				line = line.Trim();
				string[] point = line.Split(new char[] {' '});
				xyzPoints.Add (new double[] { double.Parse (point [0]), double.Parse (point [1]), double.Parse (point [2]) });
			}
		}
	
		IEnumerable<double> xs = xyzPoints.Select(e => e[0]);
		IEnumerable<double> ys = xyzPoints.Select(e => e[1]);
		IEnumerable<double> zs = xyzPoints.Select(e => e[2]);
	
		double xMin = xs.Aggregate((a, b) => a < b ? a : b);
		double yMin = ys.Aggregate((a, b) => a < b ? a : b);
		double zMin = zs.Aggregate((a, b) => a < b ? a : b);
		double xMax = xs.Aggregate((a, b) => a > b ? a : b);
		double yMax = ys.Aggregate((a, b) => a > b ? a : b);
		double zMax = zs.Aggregate((a, b) => a > b ? a : b);

		// update global variables for later use
		xRange = (float)(xMax - xMin);
		yRange = (float)(yMax - yMin);
		zRange = (float)(zMax - zMin);
		minZ = (float)zMin;
		meanX = (float)(xMin + xMax) / 2;
		meanY = (float)(yMin + yMax) / 2;

	
		xyzPoints = xyzPoints.Select (p => new double[] { p [0] - meanX, p [1] - meanY, p [2] - minZ }).ToList ();

		// return to the top of the file
		file.DiscardBufferedData();
		file.BaseStream.Seek(0, SeekOrigin.Begin);

		// copy the file to <filename>Scaled.obj, this time replacing the "v x y z" lines with the new, scaled values
		string outfilename = filename.Substring(0, filename.IndexOf(".obj")) + "Scaled.obj";

		File.Delete(Application.dataPath + "/" + outfilename);
		File.Create(Application.dataPath + "/" + outfilename).Dispose();

		StreamWriter outfile = new StreamWriter (Application.dataPath + "/" + outfilename);

		int ln = 0;
		string l;

		while ((line = file.ReadLine()) != null) {

			if (line.StartsWith ("v ")) {
				l = "v " + xyzPoints [ln][0].ToString () + " " + xyzPoints [ln][1].ToString () + " " + xyzPoints [ln][2].ToString ();
				outfile.WriteLine (l);
				ln++;
			} else {
				outfile.WriteLine (line);
			}
		}

		file.Close ();
		outfile.Close ();

    }

	public static void CreateSceneFile(string name) {
		if (!SDTP.SceneFileExists ()) {
			SDTP.modelName = name;
			SDTP.meanX = meanX;
			SDTP.meanY = meanY;
			SDTP.minZ = minZ;
			SDTP.xRange = xRange;
			SDTP.yRange = yRange;
			SDTP.zRange = zRange;
			SDTP.WriteSDTPItemsToFile ();
		}
	}

    public void MakeLandscape(float minLong, float maxLong, float minLat, float maxLat, string modelFilename, string imageFilename)
    {
		elevationData = new List<int>();
		FetchImageData(minLong, maxLong, minLat, maxLat, imageFilename);
		FetchElevationData(minLong, maxLong, minLat, maxLat);
		List<Vector3> points = ElevationPointsToXYZ(minLong, maxLong, minLat, maxLat);
		WritePointsToObj(points, modelFilename);
		CreateSceneFile (modelFilename);
    }

	public void MakeLandScapeFromRAW(string rawFile, int numRows, int numCols, float xOrigin, float yOrigin, float resolution, bool use16Bit, string modelFilename) {
		elevationData = new List<int>();
		GetElevationDataFromRaw (rawFile, numRows, numCols, use16Bit);
		List<Vector3> points = ElevationPointsToXYZ (xOrigin, yOrigin, resolution);
		WritePointsToObj(points, modelFilename);
		CreateSceneFile (modelFilename);
	}

    public void ConvertPhotogrammetryModel(string photogrammetryFilename)
    {
        ScaleDownObj (photogrammetryFilename);
		CreateSceneFile (photogrammetryFilename);
    }


    public void LandscapePhotogrammetryModel(float longRange, float latRange, string zone, string photogrammetryFilename, string modelFilename, string imageFilename)
    {
		
        Vector2 centroid = FindCentroidOfObj(photogrammetryFilename, zone);
        float lon = centroid.x;
        float lat = centroid.y;
		float minLong = (float)(Math.Round ((double)(lon - longRange / 2), 2));
		float maxLong = (float)(Math.Round ((double)(lon + longRange / 2), 2));
		float minLat = (float)(Math.Round ((double)(lat - latRange / 2), 2));
		float maxLat = (float)(Math.Round ((double)(lat + latRange / 2), 2));

		elevationData = new List<int>();
        FetchElevationData(minLong, maxLong, minLat, maxLat);
        FetchImageData(minLong, maxLong, minLat, maxLat, imageFilename);
		List<Vector3> points = ElevationPointsToXYZ(minLong, maxLong, minLat, maxLat);
		WritePointsToObj (points, modelFilename);
		CreateSceneFile (photogrammetryFilename);
    }
}
